dandiapi/api/views/version.py

```diff
@@ -102,10 +103,27 @@ def update(self, request, **kwargs):
         responses={200: VersionSerializer()},
     )
     @action(detail=True, methods=['POST'])
+    @transaction.atomic
     @method_decorator(permission_required_or_403('owner', (Dandiset, 'pk', 'dandiset__pk')))
     def publish(self, request, **kwargs):
         """Publish a version."""
-        old_version: Version = self.get_object()
+        # Note: instead of using self.get_object() here, retrieve the Version/perform permission
+        # checks manually. This is so we can use select_for_update to lock rows to prevent
+        # double publishing
+        queryset: QuerySet[Version] = self.get_queryset()
+        try:
+            old_version: Version = get_object_or_404(
+                # `nowait` is used here so that this request aborts and returns a 400 error as soon
+                # as it sees that the target Version is locked. The default behavior of
+                # `select_for_update` is to block until the lock can be acquired, but in this case
+                # we don't need to do that and incur the performance hit that comes with it.
+                queryset.select_for_update(nowait=True),
+                **kwargs,
+            )
+        except DatabaseError:
+            raise ValidationError('Dandiset is already in the process of being published')
```

Comment:
Should this really be a `ValidationError`?

Also, why is this needed, if `select_for_update()` with `nowait` results in a 404?

Final question: if a 404 arises, it should be converted into some other 400-series error specific to "already being published" right?

Reply:
> Also, why is this needed, if `select_for_update()` with `nowait` results in a 404?

`select_for_update()` with `nowait` doesn't result in a 404, it throws a `DatabaseError` if the row is locked. Then, from there I catch the `DatabaseError` and raise a `ValidationError` which does raise a 404. The original comment didn't explain this well, so I rewrote it.

> Should this really be a ValidationError?

DRF provides `ValidationError` and `ParseError` exceptions as a conveniences as opposed to manually returning a `Response` with status code. I see what you mean though, I changed it to explicitly return a `Response` with a 423 error instead.

> Final question: if a 404 arises, it should be converted into some other 400-series error specific to "already being published" right?

Yes, I changed it to a 423 (Resource is locked) error.

Reply:
Perfect!